osu.Game/Rulesets/Mods/ModMuted.cs

```diff
@@ -22,27 +27,82 @@ public abstract class ModMuted : Mod
+        private readonly BindableNumber<double> metronomeVolumeAdjust = new BindableDouble(0.5);
 
-        [SettingSource("Enable metronome", "Add a metronome to help you keep track of the rhythm.")]
+        private BindableNumber<int> currentCombo;
+
+        private AudioContainer metronomeContainer;
+
+        [SettingSource("Enable metronome", "Add a metronome beat to help you keep track of the rhythm.")]
         public BindableBool EnableMetronome { get; } = new BindableBool
         {
             Default = true,
             Value = true
         };
 
+        [SettingSource("Final volume at combo", "The combo count at which point the track reaches its final volume.")]
+        public BindableInt MuteComboCount { get; } = new BindableInt
+        {
+            Default = 100,
+            Value = 100,
+            MinValue = 0,
+            MaxValue = 500,
+        };
+
+        [SettingSource("Start muted", "Increase volume as combo builds.")]
+        public BindableBool InverseMuting { get; } = new BindableBool
+        {
+            Default = false,
+            Value = false
+        };
+
+        [SettingSource("Mute hit sounds", "Hit sounds are also muted alongside the track.")]
+        public BindableBool AffectsHitSounds { get; } = new BindableBool
+        {
+            Default = true,
+            Value = true
+        };
+
         public void ApplyToTrack(ITrack track)
         {
-            track.AddAdjustment(AdjustableProperty.Volume, volumeAdjust);
+            track.AddAdjustment(AdjustableProperty.Volume, mainVolumeAdjust);
         }
 
         public void ApplyToDrawableRuleset(DrawableRuleset<TObject> drawableRuleset)
         {
             if (EnableMetronome.Value)
-                drawableRuleset.Overlays.Add(new Metronome(drawableRuleset.Beatmap.HitObjects.First().StartTime));
+            {
+                drawableRuleset.Overlays.Add(metronomeContainer = new AudioContainer
+                {
+                    Child = new Metronome(drawableRuleset.Beatmap.HitObjects.First().StartTime)
+                });
+
+                metronomeContainer.AddAdjustment(AdjustableProperty.Volume, metronomeVolumeAdjust);
+            }
+
+            if (AffectsHitSounds.Value)
+                drawableRuleset.AudioContainer.AddAdjustment(AdjustableProperty.Volume, mainVolumeAdjust);
+        }
+
+        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            currentCombo = scoreProcessor.Combo.GetBoundCopy();
+            currentCombo.BindValueChanged(combo =>
+            {
+                double dimFactor = Math.Min(1, (double)combo.NewValue / MuteComboCount.Value);
+
+                if (InverseMuting.Value)
+                    dimFactor = 1 - dimFactor;
+
+                scoreProcessor.TransformBindableTo(metronomeVolumeAdjust, dimFactor, 500, Easing.OutQuint);
+                scoreProcessor.TransformBindableTo(mainVolumeAdjust, 1 - dimFactor, 500, Easing.OutQuint);
```

Comment:
This is probably the biggest head-turner I have right now - this may end up allocating a fair few transforms, but I'm unsure if worth fixing. Would profile but I can't on linux and still need to go on windows to do that I think...

Reply:
Yeah, you raise a good point. It would be two transforms per combo increase. I'm not sure if this is high enough to worry about, though, since `DrawableHitObject`s are probably doing many more for the same scenario.